pkg/controllers/provisioning/scheduling/nodeset.go

```diff
@@ -76,3 +102,583 @@ func (s *NodeSet) getDaemons(ctx context.Context, constraints *v1alpha5.Constrai
+				Weight:        v.Weight,
+				TopologyKey:   v.PodAffinityTerm.TopologyKey,
+				Namespaces:    buildNamespaceList(ctx, s.kubeClient, p, v.PodAffinityTerm.Namespaces, v.PodAffinityTerm.NamespaceSelector),
+				LabelSelector: v.PodAffinityTerm.LabelSelector}
+			keys = append(keys, key)
+		}
+	}
+
+	if pod.HasPodAntiAffinity(p) {
+		if err := s.trackExistingPodAntiAffinityTopology(ctx, nil, p); err != nil {
+			return err
+		}
+
+		for _, v := range p.Spec.Affinity.PodAntiAffinity.RequiredDuringSchedulingIgnoredDuringExecution {
+			key := affinityKey{
+				AntiAffinity:  true,
+				TopologyKey:   v.TopologyKey,
+				Namespaces:    buildNamespaceList(ctx, s.kubeClient, p, v.Namespaces, v.NamespaceSelector),
+				LabelSelector: v.LabelSelector}
+			keys = append(keys, key)
+		}
+		for _, v := range p.Spec.Affinity.PodAntiAffinity.PreferredDuringSchedulingIgnoredDuringExecution {
+			key := affinityKey{
+				AntiAffinity:  true,
+				Preferred:     true,
+				Weight:        v.Weight,
+				TopologyKey:   v.PodAffinityTerm.TopologyKey,
+				Namespaces:    buildNamespaceList(ctx, s.kubeClient, p, v.PodAffinityTerm.Namespaces, v.PodAffinityTerm.NamespaceSelector),
+				LabelSelector: v.PodAffinityTerm.LabelSelector}
+			keys = append(keys, key)
+		}
+	}
+
+	for _, key := range keys {
+		hash, err := hashstructure.Hash(key, hashstructure.FormatV2, &hashstructure.HashOptions{SlicesAsSets: true})
+		if err != nil {
+			return fmt.Errorf("hashing topology constraint: %w", err)
+		}
+		tsc, found := s.topologies[hash]
+		if !found {
+			var cs v1.TopologySpreadConstraint
+			cs.TopologyKey = key.TopologyKey
+			cs.LabelSelector = key.LabelSelector
+			if key.Preferred {
+				cs.WhenUnsatisfiable = v1.ScheduleAnyway
+			} else {
+				cs.WhenUnsatisfiable = v1.DoNotSchedule
+			}
+
+			// MaxSkew is unused by the affinity constraint itself, but it's useful to set it to a large value.  This
+			// is because we have to commit to topology decisions for the target pods of a pod-affinity.  If nothing
+			// else affects it, we don't really care how many of these pods end up per domain.  We only care about what
+			// the affinity owner does in relation to those pods and the domain.
+			cs.MaxSkew = math.MaxInt32
+
+			topology, err := s.buildAndUpdateTopology(ctx, cs, key.Namespaces...)
+			if err != nil {
+				return err
+			}
+			if key.AntiAffinity {
+				topology.Type = TopologyTypePodAntiAffinity
+			} else {
+				topology.Type = TopologyTypePodAffinity
+			}
+			topology.Weight = key.Weight
+
+			tsc = topology
+			s.topologies[hash] = topology
+		}
+		tsc.AddOwner(client.ObjectKeyFromObject(p))
+	}
+
+	return nil
+}
+
+// buildNamespaceList constructs a unique list of namespaces consisting of the pod's namespace and the optional list of
+// namespaces and those selected by the namespace selector
+func buildNamespaceList(ctx context.Context, c client.Client, p *v1.Pod, namespaces []string, selector *metav1.LabelSelector) []string {
```

Comment:
Let's find a way to associate this behavior with a concept. Perhaps as a member of nodeset?